tokio/src/runtime/time/mod.rs

```diff
@@ -319,23 +319,54 @@ impl Handle {
             now = lock.elapsed();
         }
 
-        while let Some(entry) = lock.poll(now) {
-            debug_assert!(unsafe { entry.is_pending() });
-
-            // SAFETY: We hold the driver lock, and just removed the entry from any linked lists.
-            if let Some(waker) = unsafe { entry.fire(Ok(())) } {
-                waker_list.push(waker);
-
-                if !waker_list.can_push() {
-                    // Wake a batch of wakers. To avoid deadlock, we must do this with the lock temporarily dropped.
-                    drop(lock);
-
-                    waker_list.wake_all();
-
-                    lock = self.inner.lock_sharded_wheel(id);
+        while let Some(expiration) = lock.poll(now) {
+            // Gets the number of entries in this slot, which will be
+            // the maximum number of times we traverse.
+            // Limiting the maximum number of iterations is very important,
+            // because it's possible that those entries might need to be
+            // reinserted into the same slot.
+
+            // This happens on the highest level, when an entry is inserted
+            // more than MAX_DURATION into the future. When this happens, we wrap
+            // around, and process some entries a multiple of MAX_DURATION before
+            // they actually need to be dropped down a level. We then reinsert them
+            // back into the same position; we must make sure we don't then process
+            // those entries again or we'll end up in an infinite loop.
+
+            // This can also happens when other threads concurrently add entries to
+            // this currently traversing slot.
+            let count = lock.get_entries_count(&expiration);
+            for _ in 0..count {
```

Comment:
I hope I understand you correctly.

What I want to address is to avoid moving all entries in the list to another list at once.

I currently see `GuardedLinkedList ` used in broadcast and notify, and it seems that it must remove all entries from the original linked list at once through calling `std::mem::take`.

Using the `std::mem::take`, we have no way to address the goal of traversing the Linked List only once. Because once the TimerHandle is removed from the original linked list, other threads need to know that it is no longer in the original linked list (they can be dropped concurrently). Here, synchronization must be achieved through locks. After modifying the state of the TimerHandle, we can not wake it up immediately because the lock is held at this time, and we need to avoid deadlock.

In short,

- Because all entries are moved at once through `std::mem::take`, we must traverse the entire list and mark them as no longer in the original linked list.
- And wakeup does not allow holding locks, so we must temporarily store these entries in the pending linked list.



The master version is:

- Get the lock
- Remove all entries from the slot linked list at once through calling `std::mem::take`.
- Traverse the slot linked list, mark all entries as `STATE_PENDING_FIRE`.
- Release the lock
- After that, other threads can know their `TimerHanle`s are stored in the pending linked list instead of the original one.
- Traverse the pending linked list in batch:
  - Get the lock
  - Traverse the pending linked list, mark some entries as `STATE_DEREGISTERED` in batch, and get their waker (32 in batch).
  - Release the lock
  - After that, other threads can know their `TimerHanle`s are not in the pending linked list or the slot linked list.
  - Wake them up.
  - Get the lock again ...

The current Version is:

Traverse the slot linked list in batch:

- Get the lock
- Traverse the slot linked list, mark some entries as `STATE_FIRING` and then `STATE_DEREGISTERED` in batch, and get their waker (32 in batch).
- Release the lock
- After that, other threads can know their `TimerHanle`s are not in the slot linked list.
- Wake them up.
- Get the lock again ...